File: ui/src/lib/topology.test.ts
import { describe, expect, it } from "vitest";
import type { TopologyGraph, TopologyTraffic } from "./api";
import {
  assemble,
  flowLabel,
  flowSeconds,
  flowTone,
  flowWidth,
  flowsByEdge,
  layout,
  lineage,
  quietBindings,
  rate,
  subtitle,
} from "./topology";

const graph: TopologyGraph = {
  nodes: [
    { id: "internet", kind: "internet", name: "Internet" },
    { id: "project/shop", kind: "project", name: "shop", project: "shop" },
    { id: "project/pricing", kind: "project", name: "pricing", project: "pricing", foreign: true },
    {
      id: "environment/shop-production",
      kind: "environment",
      name: "shop-production",
      project: "shop",
      type: "production",
      phase: "Live",
    },
    {
      id: "environment/shop-pr-7",
      kind: "environment",
      name: "shop-pr-7",
      project: "shop",
      type: "preview",
      phase: "Live",
    },
    { id: "resource/shop-db", kind: "resource", name: "shop-db", project: "shop", type: "postgres", phase: "Bound" },
    { id: "provider/neon", kind: "provider", name: "neon", type: "neon" },
    { id: "offering/pricing/api", kind: "offering", name: "api", project: "pricing", foreign: true, type: "http" },
  ],
  edges: [
    {
      id: "routes:internet->environment/shop-production",
      kind: "routes",
      from: "internet",
      to: "environment/shop-production",
    },
    { id: "routes:internet->environment/shop-pr-7", kind: "routes", from: "internet", to: "environment/shop-pr-7" },
    {
      id: "uses:environment/shop-production->resource/shop-db",
      kind: "uses",
      from: "environment/shop-production",
      to: "resource/shop-db",
    },
    {
      id: "uses:environment/shop-pr-7->resource/shop-db",
      kind: "uses",
      from: "environment/shop-pr-7",
      to: "resource/shop-db",
    },
    {
      id: "providedBy:resource/shop-db->provider/neon",
      kind: "providedBy",
      from: "resource/shop-db",
      to: "provider/neon",
    },
    {
      id: "consumes:environment/shop-production->offering/pricing/api",
      kind: "consumes",
      from: "environment/shop-production",
      to: "offering/pricing/api",
    },
    {
      id: "serves:offering/pricing/api->project/pricing",
      kind: "serves",
      from: "offering/pricing/api",
      to: "project/pricing",
    },
  ],
};

const traffic: TopologyTraffic = {
  since: "2026-09-25T10:00:00Z",
  until: "2026-09-25T10:05:00Z",
  nodes: [
    { id: "project/blog", kind: "project", name: "blog", project: "blog", foreign: true },
    { id: "external/api.stripe.com", kind: "external", name: "api.stripe.com" },
  ],
  edges: [
    {
      from: "internet",
      to: "environment/shop-production",
      protocol: "HTTP",
      flows: 600,
      rps: 2,
      errors: 0,
      drops: 0,
      p95Ms: 40,
      status: "declared",
      along: ["routes:internet->environment/shop-production"],
    },
    {
      from: "project/blog",
      to: "environment/shop-production",
      protocol: "HTTP",
      flows: 10,
      rps: 0.03,
      errors: 2,
      drops: 0,
      p95Ms: 12,
      status: "undeclared",
    },
  ],
};

describe("assemble", () => {
  it("leaves previews out unless asked, with their edges", () => {
    const diagram = assemble(graph, null, { previews: false });
    expect(diagram.nodes.map((n) => n.id)).not.toContain("environment/shop-pr-7");
    expect(diagram.edges.some((e) => e.from === "environment/shop-pr-7" || e.to === "environment/shop-pr-7")).toBe(
      false,
    );
    const everything = assemble(graph, null, { previews: true });
    expect(everything.nodes.map((n) => n.id)).toContain("environment/shop-pr-7");
  });

  it("draws a project as a box only when an edge needs it", () => {
    const ids = assemble(graph, null, { previews: false }).nodes.map((n) => n.id);
    // shop has boxes of its own, so its name is the group, not a box
    expect(ids).not.toContain("project/shop");
    // pricing is foreign and answers an offering, so it is a box
    expect(ids).toContain("project/pricing");
  });

  it("adds the observed pairs nothing declares, and the boxes only they reach", () => {
    const diagram = assemble(graph, traffic, { previews: false });
    const observed = diagram.edges.filter((e) => e.observed);
    expect(observed.map((e) => e.from)).toEqual(["project/blog"]);
    const ids = diagram.nodes.map((n) => n.id);
    expect(ids).toContain("project/blog");
    // reached by nothing in this reading, so not drawn
    expect(ids).not.toContain("external/api.stripe.com");
  });

  it("draws a declared pair on its own when a focus left its declared edges out", () => {
    const focused = { nodes: graph.nodes, edges: graph.edges.filter((e) => e.kind !== "routes") };
    const diagram = assemble(focused, traffic, { previews: false });
    expect(diagram.edges.some((e) => e.observed?.status === "declared" && e.from === "internet")).toBe(true);
    // and not when the edges are there to carry it
    const whole = assemble(graph, traffic, { previews: false });
    expect(whole.edges.some((e) => e.observed?.status === "declared")).toBe(false);
  });
});

describe("layout", () => {
  it("draws every edge left to right when there is no cycle", () => {
    const diagram = assemble(graph, traffic, { previews: true });
    const { placed } = layout(diagram.nodes, diagram.edges);
    for (const edge of diagram.edges) {
      const from = placed.get(edge.from)!;
      const to = placed.get(edge.to)!;
      expect(from.layer, `${edge.from} -> ${edge.to}`).toBeLessThan(to.layer);
    }
  });

  it("puts a source right in front of what it points at", () => {
    const diagram = assemble(graph, traffic, { previews: false });
    const { placed } = layout(diagram.nodes, diagram.edges);
    expect(placed.get("project/blog")!.layer).toBe(placed.get("environment/shop-production")!.layer - 1);
  });

  it("survives a cycle", () => {
    const nodes = [
      { id: "a", kind: "environment" as const, name: "a" },
      { id: "b", kind: "environment" as const, name: "b" },
    ];
    const { placed } = layout(nodes, [
      { from: "a", to: "b" },
      { from: "b", to: "a" },
    ]);
    expect(placed.size).toBe(2);
    expect(placed.get("a")!.layer).not.toBe(placed.get("b")!.layer);
  });

  it("never stacks two boxes on each other", () => {
    const diagram = assemble(graph, traffic, { previews: true });
    const boxes = [...layout(diagram.nodes, diagram.edges).placed.values()];
    const spots = new Set(boxes.map((b) => `${b.x},${b.y}`));
    expect(spots.size).toBe(boxes.length);
  });
});

describe("flows", () => {
  it("credits a pair to every declared edge it runs along", () => {
    const flows = flowsByEdge(traffic);
    expect(flows.get("routes:internet->environment/shop-production")?.rps).toBe(2);
    expect(flows.size).toBe(1);
  });

  it("finds the bindings nothing used", () => {
    const quiet = quietBindings(graph, flowsByEdge(traffic));
    expect(quiet.map((e) => e.id)).toEqual(["consumes:environment/shop-production->offering/pricing/api"]);
  });

  it("moves busier edges faster and draws them thicker", () => {
    expect(flowSeconds(100)).toBeLessThan(flowSeconds(1));
    expect(flowSeconds(0)).toBeLessThanOrEqual(3);
    expect(flowWidth(100)).toBeGreaterThan(flowWidth(1));
  });

  it("colours by the share of failures, not their count", () => {
    const base = { rps: 1, p95Ms: 0, protocol: "HTTP", drops: 0 };
    expect(flowTone({ ...base, flows: 1000, errors: 1 })).toBe("warning");
    expect(flowTone({ ...base, flows: 10, errors: 2 })).toBe("error");
    expect(flowTone({ ...base, flows: 10, errors: 0 })).toBe("primary");
  });

  it("says a rate in few characters", () => {
    expect(rate(0)).toBe("0/s");
    expect(rate(0.001)).toBe("<0.01/s");
    expect(rate(0.25)).toBe("0.25/s");
    expect(rate(3.14)).toBe("3.1/s");
    expect(rate(42.4)).toBe("42/s");
    expect(flowLabel({ rps: 2, flows: 10, errors: 1, drops: 0, p95Ms: 40.2, protocol: "HTTP" })).toBe(
      "2.0/s · p95 40ms · 1 5xx",
    );
  });
});

describe("lineage", () => {
  it("is everything upstream and downstream, and nothing beside", () => {
    const seen = lineage("resource/shop-db", graph.edges);
    expect(seen.has("provider/neon")).toBe(true);
    expect(seen.has("environment/shop-production")).toBe(true);
    expect(seen.has("internet")).toBe(true);
    // shop-production's *other* dependency is not upstream of the database
    expect(seen.has("offering/pricing/api")).toBe(false);
  });
});

describe("subtitle", () => {
  it("names what a box is without the cluster's words", () => {
    expect(subtitle(graph.nodes[3])).toBe("production · Live");
    expect(subtitle(graph.nodes[5])).toBe("postgres");
    expect(subtitle(graph.nodes[2])).toBe("another team's project");
  });
});

File: internal/api/topology_test.go
/*
Copyright 2026.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package api

import (
	"net/http"
	"testing"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"

	kitchenv1alpha1 "github.com/Bermos/Kitchen/api/v1alpha1"
	"github.com/Bermos/Kitchen/internal/clickhouse"
	"github.com/Bermos/Kitchen/internal/controller"
)

// The architecture overview: the declared graph read off the objects, and the
// flow collector's edges laid over it.

const (
	pricingEnvironment = "pricing-production"
	cacheInstance      = "kitchen-shop-cache-1a2b"
	httpProtocol       = "HTTP"
)

// topologyFixtures are the shop fixtures plus a second project whose offering
// shop's production binds, a cache behind the shop, and a worker process.
func topologyFixtures(phase kitchenv1alpha1.ClaimPhase) []runtime.Object {
	provider := pricingProject(openOffering())
	providerEnv := &kitchenv1alpha1.Environment{
		ObjectMeta: metav1.ObjectMeta{Name: pricingEnvironment, Namespace: testNamespace},
		Spec: kitchenv1alpha1.EnvironmentSpec{
			ProjectRef: kitchenv1alpha1.LocalObjectReference{Name: providerProject},
			Type:       kitchenv1alpha1.EnvironmentProduction,
		},
		Status: kitchenv1alpha1.EnvironmentStatus{Phase: kitchenv1alpha1.EnvironmentLive},
	}
	binding := &kitchenv1alpha1.ResourceClaim{
		ObjectMeta: metav1.ObjectMeta{Name: "shop-pricing", Namespace: testNamespace},
		Spec: kitchenv1alpha1.ResourceClaimSpec{
			ProjectRef: kitchenv1alpha1.LocalObjectReference{Name: feedProject},
			Type:       kitchenv1alpha1.ClaimTypeService,
			Config: &runtime.RawExtension{
				Raw: []byte(`{"service": {"project": "pricing", "offering": "pricing-api"}}`),
			},
		},
		Status: kitchenv1alpha1.ResourceClaimStatus{
			Phase: phase,
			Service: &kitchenv1alpha1.ClaimServiceStatus{Bindings: []kitchenv1alpha1.ClaimServiceBinding{
				{Consumer: kitchenv1alpha1.EnvironmentProduction, Environment: pricingEnvironment},
			}},
		},
	}
	cache := &kitchenv1alpha1.ResourceClaim{
		ObjectMeta: metav1.ObjectMeta{Name: "shop-cache", Namespace: testNamespace},
		Spec: kitchenv1alpha1.ResourceClaimSpec{
			ProjectRef:    kitchenv1alpha1.LocalObjectReference{Name: feedProject},
			ConnectionRef: &kitchenv1alpha1.LocalObjectReference{Name: "valkey"},
			Type:          kitchenv1alpha1.ClaimTypeRedis,
		},
		Status: kitchenv1alpha1.ResourceClaimStatus{Phase: kitchenv1alpha1.ClaimBound, InstanceID: cacheInstance},
	}
	objects := []runtime.Object{provider, providerEnv, binding, cache}
	return append(objects, fixtures()...)
}

func nodesByID(nodes []topologyNode) map[string]topologyNode {
	out := make(map[string]topologyNode, len(nodes))
	for _, node := range nodes {
		out[node.ID] = node
	}
	return out
}

func edgesByID(edges []topologyEdge) map[string]topologyEdge {
	out := make(map[string]topologyEdge, len(edges))
	for _, edge := range edges {
		out[edge.ID] = edge
	}
	return out
}

func edgeID(kind, from, to string) string { return kind + ":" + from + "->" + to }

func TestTheTopologyIsReadOffTheObjects(t *testing.T) {
	h := newHarness(t, nil, topologyFixtures(kitchenv1alpha1.ClaimBound)...)

	res := h.do(t, http.MethodGet, "/api/v1/topology", "")
	if res.Code != http.StatusOK {
		t.Fatalf("GET /topology = %d: %s", res.Code, res.Body.String())
	}
	graph := decode[topologyGraph](t, res)
	nodes := nodesByID(graph.Nodes)
	edges := edgesByID(graph.Edges)

	shopEnv := environmentNodeID(testEnvironment)
	offering := offeringNodeID(providerProject, openOfferingName)
	for _, id := range []string{
		projectNodeID(feedProject), projectNodeID(providerProject), shopEnv, environmentNodeID(pricingEnvironment),
		domainNodeID("shop.example.com"), topologyInternet, resourceNodeID("shop-db"), providerNodeID("neon"),
		resourceNodeID("shop-cache"), offering,
	} {
		if _, ok := nodes[id]; !ok {
			t.Errorf("the graph should have %s, got %v", id, graph.Nodes)
		}
	}
	if env := nodes[shopEnv]; env.Project != feedProject || env.Type != string(kitchenv1alpha1.EnvironmentProduction) || env.URL == "" {
		t.Errorf("an environment carries its project, type and address: %+v", env)
	}
	if node := nodes[offering]; node.Foreign || node.Type != "http" || node.Detail != "open" {
		t.Errorf("an operator's offering is nobody's foreign one and says what it speaks: %+v", node)
	}

	for _, id := range []string{
		edgeID(topologyRoutes, topologyInternet, shopEnv),
		edgeID(topologyRoutes, topologyInternet, domainNodeID("shop.example.com")),
		edgeID(topologyRoutes, domainNodeID("shop.example.com"), shopEnv),
		edgeID(topologyUses, shopEnv, resourceNodeID("shop-db")),
		edgeID(topologyProvidedBy, resourceNodeID("shop-db"), providerNodeID("neon")),
		edgeID(topologyConsumes, shopEnv, offering),
		edgeID(topologyServes, offering, environmentNodeID(pricingEnvironment)),
	} {
		edge, ok := edges[id]
		if !ok {
			t.Errorf("the graph should have edge %s, got %v", id, graph.Edges)
			continue
		}
		if edge.State != "" {
			t.Errorf("a bound edge is in effect, got state %q on %s", edge.State, id)
		}
	}
	// The binding reaches only production, so it is the only class drawn;
	// nothing about a service claim is a resource of the consumer's.
	if _, ok := nodes[resourceNodeID("shop-pricing")]; ok {
		t.Errorf("a service claim is an edge to an offering, not a resource")
	}
}

// A member sees their own project whole, and the other end of a binding as a
// name: no environment of somebody else's, and nothing of a project that
// nothing of theirs names.
func TestTheTopologyNarrowsToTheCallersProjects(t *testing.T) {
	h := asMember(t, kitchenv1alpha1.AccessRoleViewer,
		append(topologyFixtures(kitchenv1alpha1.ClaimBound)[:4], blogFixtures()...)...)

	res := h.do(t, http.MethodGet, "/api/v1/topology", "")
	if res.Code != http.StatusOK {
		t.Fatalf("GET /topology = %d: %s", res.Code, res.Body.String())
	}
	graph := decode[topologyGraph](t, res)
	nodes := nodesByID(graph.Nodes)

	if _, ok := nodes[environmentNodeID(pricingEnvironment)]; ok {
		t.Errorf("another project's environment is not the member's to see")
	}
	for _, id := range []string{projectNodeID(otherProject), environmentNodeID("blog-production")} {
		if _, ok := nodes[id]; ok {
			t.Errorf("a project nothing of the member's names is not on their graph: %s", id)
		}
	}
	provider, ok := nodes[projectNodeID(providerProject)]
	if !ok || !provider.Foreign {
		t.Fatalf("the provider of a bound offering is drawn as a foreign name, got %+v", provider)
	}
	offering := offeringNodeID(providerProject, openOfferingName)
	if node := nodes[offering]; !node.Foreign {
		t.Errorf("an offering of a project the member cannot see is foreign: %+v", node)
	}
	edges := edgesByID(graph.Edges)
	if _, ok := edges[edgeID(topologyServes, offering, projectNodeID(providerProject))]; !ok {
		t.Errorf("a foreign offering is served by its project as a whole, got %v", graph.Edges)
	}
}

// The provider's side of the same binding: who depends on me, as a name.
func TestTheTopologyNamesTheConsumersOfTheCallersOfferings(t *testing.T) {
	h := newHarness(t, nil, topologyFixtures(kitchenv1alpha1.ClaimBound)...)
	h.demoteCaller(t)
	h.grant(t, providerProject, kitchenv1alpha1.AccessRoleViewer)

	graph := decode[topologyGraph](t, h.do(t, http.MethodGet, "/api/v1/topology", ""))
	nodes := nodesByID(graph.Nodes)
	if consumer := nodes[projectNodeID(feedProject)]; !consumer.Foreign {
		t.Errorf("the consumer is a foreign name to the provider, got %+v", consumer)
	}
	if _, ok := nodes[resourceNodeID("shop-db")]; ok {
		t.Errorf("the consumer's own resources are not the provider's to see")
	}
	edges := edgesByID(graph.Edges)
	offering := offeringNodeID(providerProject, openOfferingName)
	if _, ok := edges[edgeID(topologyConsumes, projectNodeID(feedProject), offering)]; !ok {
		t.Errorf("the consumer binds the offering as a project, got %v", graph.Edges)
	}
}

func TestATopologyFocusKeepsOneProjectAndItsNeighbours(t *testing.T) {
	h := newHarness(t, nil, topologyFixtures(kitchenv1alpha1.ClaimBound)...)

	graph := decode[topologyGraph](t, h.do(t, http.MethodGet, "/api/v1/topology?project=pricing", ""))
	nodes := nodesByID(graph.Nodes)
	if _, ok := nodes[environmentNodeID(testEnvironment)]; !ok {
		t.Errorf("the consumer of the focused project's offering is one edge away and stays")
	}
	if _, ok := nodes[resourceNodeID("shop-db")]; ok {
		t.Errorf("the consumer's own database is two edges away and goes")
	}
}

func TestAPendingBindingIsDrawnAsWaiting(t *testing.T) {
	h := newHarness(t, nil, topologyFixtures(kitchenv1alpha1.ClaimPendingApproval)...)

	graph := decode[topologyGraph](t, h.do(t, http.MethodGet, "/api/v1/topology", ""))
	edge := edgesByID(graph.Edges)[edgeID(topologyConsumes, environmentNodeID(testEnvironment),
		offeringNodeID(providerProject, openOfferingName))]
	if edge.State != string(kitchenv1alpha1.ClaimPendingApproval) {
		t.Errorf("a binding waiting on the provider says so, got %+v", edge)
	}
}

func TestObservedTrafficIsAttributedOntoTheGraph(t *testing.T) {
	h := newHarness(t, nil, append(topologyFixtures(kitchenv1alpha1.ClaimBound), blogFixtures()...)...)
	shop := controller.AppNamespace(feedProject)
	h.logs.edges = []clickhouse.TrafficEdge{
		// Visitors, through a published address.
		{Source: worldEndpoint, Destination: testEnvironment, DestinationNamespace: shop,
			Protocol: httpProtocol, Flows: 600, RPS: 10},
		// The binding, used.
		{Source: testEnvironment, SourceNamespace: shop, Destination: pricingEnvironment,
			DestinationNamespace: controller.AppNamespace(providerProject), Protocol: httpProtocol, Flows: 60, RPS: 1},
		// The cache, by the name its provider ran it under.
		{Source: testEnvironment + "-worker", SourceNamespace: shop, Destination: cacheInstance + "-0",
			DestinationNamespace: "kitchen-caches", Protocol: "TCP", Flows: 30, RPS: 0.5},
		// Somebody calling the shop with nothing saying they may.
		{Source: "blog-production", SourceNamespace: controller.AppNamespace(otherProject),
			Destination: testEnvironment, DestinationNamespace: shop, Protocol: httpProtocol, Flows: 6, RPS: 0.1,
			Errors: 2},
		// Off the platform.
		{Source: testEnvironment, SourceNamespace: shop, Destination: "api.stripe.com",
			Protocol: httpProtocol, Flows: 3, RPS: 0.05},
		// The platform's own components.
		{Source: "preview-gate", SourceNamespace: testNamespace, Destination: testEnvironment,
			DestinationNamespace: shop, Protocol: httpProtocol, Flows: 2},
	}

	res := h.do(t, http.MethodGet, "/api/v1/topology/traffic", "")
	if res.Code != http.StatusOK {
		t.Fatalf("GET /topology/traffic = %d: %s", res.Code, res.Body.String())
	}
	answer := decode[topologyTraffic](t, res)
	if answer.Since.IsZero() || !answer.Until.After(answer.Since) {
		t.Errorf("the answer says which window its rates are over, got %v to %v", answer.Since, answer.Until)
	}

	shopEnv := environmentNodeID(testEnvironment)
	offering := offeringNodeID(providerProject, openOfferingName)
	byPair := map[string]observedEdge{}
	for _, edge := range answer.Edges {
		byPair[edge.From+" "+edge.To] = edge
	}
	for pair, want := range map[string]struct {
		status string
		along  []string
	}{
		topologyInternet + " " + shopEnv: {observedDeclared,
			[]string{edgeID(topologyRoutes, topologyInternet, shopEnv)}},
		shopEnv + " " + environmentNodeID(pricingEnvironment): {observedDeclared, []string{
			edgeID(topologyConsumes, shopEnv, offering),
			edgeID(topologyServes, offering, environmentNodeID(pricingEnvironment)),
		}},
		shopEnv + " " + resourceNodeID("shop-cache"): {observedDeclared,
			[]string{edgeID(topologyUses, shopEnv, resourceNodeID("shop-cache"))}},
		environmentNodeID("blog-production") + " " + shopEnv: {observedUndeclared, nil},
		shopEnv + " " + externalNodeID("api.stripe.com"):     {observedExternal, nil},
		topologyPlatform + " " + shopEnv:                     {observedPlatform, nil},
	} {
		edge, ok := byPair[pair]
		if !ok {
			t.Errorf("no observed edge %s, got %+v", pair, answer.Edges)
			continue
		}
		if edge.Status != want.status || len(edge.Along) != len(want.along) {
			t.Errorf("%s = %s along %v, want %s along %v", pair, edge.Status, edge.Along, want.status, want.along)
			continue
		}
		for i := range want.along {
			if edge.Along[i] != want.along[i] {
				t.Errorf("%s runs along %v, want %v", pair, edge.Along, want.along)
			}
		}
	}

	extra := nodesByID(answer.Nodes)
	for _, id := range []string{topologyPlatform, externalNodeID("api.stripe.com")} {
		if _, ok := extra[id]; !ok {
			t.Errorf("a node only the flows reach is answered with them: %s, got %v", id, answer.Nodes)
		}
	}
	if _, ok := extra[shopEnv]; ok {
		t.Errorf("a node the declared graph already has is not repeated")
	}
}

// A member is told that somebody calls their project, and who — a project, by
// name — but not which of that project's environments.
func TestObservedTrafficNamesAnotherProjectAsAWhole(t *testing.T) {
	h := asMember(t, kitchenv1alpha1.AccessRoleViewer, blogFixtures()...)
	h.logs.edges = []clickhouse.TrafficEdge{{
		Source: "blog-production", SourceNamespace: controller.AppNamespace(otherProject),
		Destination: testEnvironment, DestinationNamespace: controller.AppNamespace(feedProject),
		Protocol: httpProtocol, Flows: 6,
	}}

	answer := decode[topologyTraffic](t, h.do(t, http.MethodGet, "/api/v1/topology/traffic", ""))
	if len(answer.Edges) != 1 || answer.Edges[0].From != projectNodeID(otherProject) {
		t.Fatalf("the caller is a project as a whole, got %+v", answer.Edges)
	}
	if node := nodesByID(answer.Nodes)[projectNodeID(otherProject)]; !node.Foreign {
		t.Errorf("a project the member cannot see is foreign, got %+v", node)
	}
}

func TestObservedTrafficAnswersANoStoreInstallationPlainly(t *testing.T) {
	h := newHarness(t, nil, fixtures()...)
	h.logs.trafficErr = errNoLogStore

	res := h.do(t, http.MethodGet, "/api/v1/topology/traffic", "")
	if res.Code == http.StatusOK {
		t.Fatalf("a failed store read is not an empty graph: %s", res.Body.String())
	}
}

// Narrowed to one project, the traffic still names both ends of every pair it
// answers with, even where the focused graph left one of them out.
func TestFocusedTrafficCarriesTheEndsTheFocusLeftOut(t *testing.T) {
	h := newHarness(t, nil, topologyFixtures(kitchenv1alpha1.ClaimBound)...)
	h.logs.edges = []clickhouse.TrafficEdge{{
		Source: testEnvironment + "-worker", SourceNamespace: controller.AppNamespace(feedProject),
		Destination: cacheInstance + "-0", DestinationNamespace: "kitchen-caches", Protocol: "TCP", Flows: 30,
	}}

	answer := decode[topologyTraffic](t, h.do(t, http.MethodGet, "/api/v1/topology/traffic?project=pricing", ""))
	nodes := nodesByID(answer.Nodes)
	if _, ok := nodes[resourceNodeID("shop-cache")]; !ok {
		t.Errorf("a resource two edges from the focus is answered with the traffic reaching it, got %v", answer.Nodes)
	}
	if _, ok := nodes[environmentNodeID(testEnvironment)]; ok {
		t.Errorf("an end the focused graph already has is not repeated")
	}
}
